src/lib/Microsoft.Fx.Portability.Reports.DGML/ReferenceNode.cs

```diff
@@ -66,65 +66,65 @@ public double GetPortabilityIndexForReferences(int target)
+            if (availableApis == 0 && unavailableApis == 0)
+                return 0;
 
-                return availableApis / ((double)availableApis + unavailableApis);
-            }
+            return availableApis / ((double)availableApis + unavailableApis);
         }
 
-        public bool TryGetAPICountFromReferences(int target, out int availAPIs, out int unavailAPIs)
+        private void GetAPICountFromReferences(int target, out int availAPIs, out int unavailAPIs)
         {
-            availAPIs = UsageData[target].GetAvailableAPICalls();
-            unavailAPIs = UsageData[target].GetUnavailableAPICalls();
-
-            // We are going to use a flag on the object to detect if we have a reference cycle while computing the APIs for the references.
-            if (_searchInGraph == true)
-            {
-                // Cycle!!!
-                _searchInGraph = false; // Reset this flag
-                return false;
-            }
-            else
-            {
-                _searchInGraph = true;
-            }
+            availAPIs = 0;
+            unavailAPIs = 0;
 
             foreach (var item in Nodes)
             {
-                if (item.UsageData == null)
+                // We are going to use the flag on the object to detect if we have run into this node in the reference graph while computing the APIs for the references,
+                // because we want to only count a reference node once if it is referenced in the graph more than once.
+                if (item._searchInGraph == true)
                 {
-                    // skip the Node with no UsageData
                     continue;
                 }
-
-                if (!item.TryGetAPICountFromReferences(target, out int refCountAvail, out int refCountUnavail))
+                else
                 {
-                    // Cycle!
-                    _searchInGraph = false; // Reset this flag
+                    item._searchInGraph = true;
+                }
 
-                    return false;
+                if (item.UsageData != null)
+                {
+                    availAPIs += item.UsageData[target].GetAvailableAPICalls();
+                    unavailAPIs += item.UsageData[target].GetUnavailableAPICalls();
                 }
 
+                item.GetAPICountFromReferences(target, out int refCountAvail, out int refCountUnavail);
+
                 availAPIs += refCountAvail;
                 unavailAPIs += refCountUnavail;
             }
+        }
+
+        /// <summary>
+        /// reset all the nodes in the reference graph as not be searched.
+        /// </summary>
+        private void ResetReferenceSearchGraph()
+        {
+            // if we don't have any outgoing references, done reset with this node
+            if (Nodes.Count == 0)
+                return;
+            foreach (var item in Nodes)
+            {
+                item._searchInGraph = false;
+                item.ResetReferenceSearchGraph();
```

Comment:
How do you avoid circular references here?

Reply:
For what it's worth, my strategy for avoiding following cycles in graphs is to put each node in a HashSet<T> and if Add returns true, process it, otherwise, I've seen it before. It works well for graphs that are reasonable in size, and it doesn't  come with any interesting cleanup problems like this. You can also have multiple threads inspecting the same graph without colliding on state.

Reply:
If you wanted a fix for this approach, ResetReferenceSearchGraph could treat _searchInGraph with the oppose rule (only follow Nodes if _searchInGraph was false).

Reply:
Great suggestion! Pushed a commit to use the suggested HashSet approach.